File: app_admin/lib/features/payment/domain/entities/payment.dart
import 'package:equatable/equatable.dart';

enum PaymentStatus { pending, paid, failed, cancelled }

class Payment extends Equatable {
  final String id;
  final String clientId;
  final String clientName;
  final double amount;
  final PaymentStatus status;
  final String? description;
  final String? method;
  final DateTime? dueDate;
  final DateTime? paidAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Payment({
    required this.id,
    required this.clientId,
    required this.clientName,
    required this.amount,
    required this.status,
    this.description,
    this.method,
    this.dueDate,
    this.paidAt,
    required this.createdAt,
    required this.updatedAt,
  });

  @override
  List<Object?> get props =>
      [id, clientId, amount, status, description, method, dueDate, paidAt];
}

File: api/internal/domain/payment/payment.go
package payment

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

type Status string

const (
	StatusPending   Status = "pending"
	StatusPaid      Status = "paid"
	StatusFailed    Status = "failed"
	StatusCancelled Status = "cancelled"
)

type Payment struct {
	ID          uuid.UUID  `db:"id"`
	ClientID    uuid.UUID  `db:"client_id"`
	Amount      float64    `db:"amount"`
	Status      Status     `db:"status"`
	Description *string    `db:"description"`
	Method      *string    `db:"method"`
	DueDate     *time.Time `db:"due_date"`
	PaidAt      *time.Time `db:"paid_at"`
	CreatedAt   time.Time  `db:"created_at"`
	UpdatedAt   time.Time  `db:"updated_at"`
}

func NewPayment(clientID uuid.UUID, amount float64) (*Payment, error) {
	if clientID == uuid.Nil {
		return nil, errors.New("client_id is required")
	}
	if amount <= 0 {
		return nil, errors.New("amount must be greater than zero")
	}
	return &Payment{
		ID:        uuid.New(),
		ClientID:  clientID,
		Amount:    amount,
		Status:    StatusPending,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}, nil
}

type WriteRepository interface {
	Save(p *Payment) error
}

// PaymentWithClient is a read model that includes client name from JOIN.
type PaymentWithClient struct {
	Payment
	ClientName string `db:"client_name"`
}

type ReadRepository interface {
	FindByID(id uuid.UUID) (*PaymentWithClient, error)
	FindAll(clientID *uuid.UUID, status *Status, offset, limit int) ([]*PaymentWithClient, int, error)
}
